### PR TITLE
ziti-edge-tunnel: Initial commit ziti-edge-tunnel transparent proxy

### PR DESCRIPTION
**Pull request by mjtrangoni**
Hi @qrkourier,

as we talked in #129 I am adding here a new chart for my use case. Also added a note about the current problem of this configuration.

Let me know if I need to adjust something. I tested it on my end, and installation/upgrade is running as it should.

### CODE REVIEW

**Comment by mjtrangoni** [comment]
@qrkourier may I help with something on CI side or is it just a time matter? Thank you!

**Comment by qrkourier** [comment]
Hi @mjtrangoni, I appreciate you sending this contribution. I consulted with @scareything to confirm my understanding of the uniqueness of Ziti's intercept IPs. They're unique to the device/node, but this solution results in pods sending DNS queries to the node where CoreDNS is running. This will cause the following failure modes:

1. The DNS response will contain the incorrect IP when multiple Ziti service addresses are available, and the K8s cluster has more than one node. This will result in the client application's traffic flowing to the wrong server application. The application will not function reliably.
2. Ziti wildcard addresses will not function because the intercept addresses are allocated dynamically, exclusively on the device/node where the DNS query is handled, i.e., the node where CoreDNS is running.

In light of the unmitigated risk of these failure modes I don't think we should continue advertising the node proxy solution for Kubernetes.

The path toward a node-level proxy probably involves some CoreDNS research. It might be possible to ensure that only CoreDNS on the same node as the pod answers DNS queries. In that case, these problems do not exist.

**Comment by qrkourier** [comment]
I will close this until we have a node proxy solution that's not vulnerable to the remote DNS problem.

**Comment by mjtrangoni** [comment]
@qrkourier I think I have a solution for this very special case, which handles the problem with DNS. 
**Disclaimer:** We use and tested it with [cilium](https://cilium.io/) as CNI, but it should also work with other CNIs like [Traefik](https://traefik.io/traefik/) as well.

Having a k8s with this Helm Chart installed, also, running on every node as DaemonSet, we proceed to install Cilium with the `Local-Redirect-Policy` enabled  and install `node-local-dns`, which works as a local cache for DNS, and also runs as DaemonSet on every node. See the example [here](https://docs.cilium.io/en/v1.13/network/kubernetes/local-redirect-policy/#node-local-dns-cache). 

Only one extra step is needed to enable the local Ziti DNS here, which is adding the forward rule to `node-local-dns` ConfigMap, as follows,
```console
   Corefile: |
+    your.openziti.domain:53 {
+        errors
+        cache 30
+        reload
+        loop
+        bind 0.0.0.0
+        forward . 100.64.0.2
+        prometheus :9253
+        log
+        }
     cluster.local:53 {
         errors
         cache {
```
Delete the node-local-dns pod to reload the **Corefile** config. Just run one debug pod, and do some DNS lookup, after that prometheus will show something like,

```
# curl -s -k 172.28.0.185:9253/metrics | grep coredns_dns_request_count
# HELP coredns_dns_request_count_total Counter of DNS requests made per zone, protocol and family.
# TYPE coredns_dns_request_count_total counter
coredns_dns_request_count_total{family="1",proto="udp",server="dns://0.0.0.0:53",zone="."} 169
coredns_dns_request_count_total{family="1",proto="udp",server="dns://0.0.0.0:53",zone="your.openziti.domain."} 7
coredns_dns_request_count_total{family="1",proto="udp",server="dns://0.0.0.0:53",zone="cluster.local."} 514
coredns_dns_request_count_total{family="1",proto="udp",server="dns://0.0.0.0:53",zone="in-addr.arpa."} 2
coredns_dns_request_count_total{family="1",proto="udp",server="dns://0.0.0.0:53",zone="ip6.arpa."} 1
```

Let me know if you agree, and we can proceed?

**Comment by qrkourier** [comment]
Do I understand correctly? This requires the cluster operator to use https://kubernetes.io/docs/tasks/administer-cluster/nodelocaldns/ which provides a CoreDNS caching nameserver on each node. Then, pods using `dnsPolicy` `ClusterFirst` will automatically try first to resolve domain names with the local nameserver IP, and subsequently falling back to the CoreDNS cluster service VIP.

With your proposed patch to the `node-local-dns` ConfigMap, queries handled by the local nameserver could take a new path, in addition to the cache path. If the query's RNAME matches the Ziti namespace then it is forwarded to the static Ziti nameserver IP on the same node, not evaluating the cache or recursing to the CoreDNS cluster service VIP at all.

The end result is that the cluster-level default `ClusterFirst` policy configures pods to query the local caching CoreDNS instance first, which are distributed one per node as a daemonset, and as long as the Ziti tunnel daemonset too is distributed one per node, those pods will be able to:

1. resolve the correct Ziti intercept IP for the same node
2. route packets to the intercept IP on the same node

If so, it sounds like this eliminates the majority of the risk of the remote DNS problem. Remaining questions include how best to orchestrate this. We could host the ziti-edge-tunnel chart as is, and provide instructions for orchestrating this as a node proxy. Eventually, a Ziti operator could optionally configure the required pair of daemonsets: `node-local-dns` and `ziti-edge-tunnel`, with the necessary patch to the default ConfigMap.

**Comment by sa-ChristianAnton** [comment]
Hi @qrkourier,

we have investigated this a bit more in detail, and I want to drop in to this discussion. You have got the functionality of node-local-dns correctly, this is exactly how it is designed to work: whatever pod tries to reach "a DNS" and its request is forwarded to the one node-local-dns pod that is running on that same node. These pods are indeed just CoreDNS pods, running as Daemonset and forwarding to the cluster DNS (the CoreDNS deployed within Kubernetes) as an upstream. The original idea of this is to cache entries locally, but its architecture fits perfectly into the Openziti use case: either by forwarding to the locally configured DNS servers (resolv.conf) or by configuring an additional configuration within these node-local-dns CoreDNS pods, you make sure the Openziti service name is resolved on the node where the pod originating the request is running.

Now, the difference in within to deploy node-local-dns with Cilium or without is just that in default setup as described [here](https://kubernetes.io/docs/tasks/administer-cluster/nodelocaldns/) it is necessary to run the node-local-dns pods with "hostNetwork: true" and use iptables to redirect the traffic directed to the cluster DNS Service address to the locally running link-local address of the node-local-dns pod, while with the Cilium method this is not necessary because Cilium has a `CiliumLocalRedirectPolicy` CRD definition (described [here](https://docs.cilium.io/en/v1.13/network/kubernetes/local-redirect-policy/).

Long story short, I would opt for a big fat remark in this Helm chart's README.md that points to the fact that this helm chart is only reliably usable in either single-node K8S clusters or, in case of multi-node in combination with the node-local-dns.

Of course, at some point more automation would be cool, but as always, the specific setup will depend a lot on how the cluster is built and therefore it should be ok to just explain this fact and show the solution.

**Comment by qrkourier** [comment]
@sa-ChristianAnton That's incredible. Thank you for spelling it out. I will get to try this myself soon.

@mjtrangoni It seems best to compile relevant changes in this branch. Excited to see this working for others too.

**Comment by mjtrangoni** [comment]
Done. Let me know if you want to extend/change the README note.

**Comment by mjtrangoni** [comment]
@qrkourier renamed

**Review by qrkourier — APPROVED** [review]
:shipit:
